internal/dag/cache.go

```diff
@@ -511,38 +511,56 @@ func (kc *KubernetesCache) secretTriggersRebuild(secret *v1.Secret) bool {
 		return true
 	}
 
-	delegations := make(map[string]bool) // targetnamespace/secretname to bool
+	secret := types.NamespacedName{
+		Namespace: secretObj.Namespace,
+		Name:      secretObj.Name,
+	}
+
+	delegatedForAnyNamespace := false // True if this secret has been delegated to any namespace.
+	delegations := []string{}         // List of namespaces that this secret has been delegated to.
 
-	// TODO(youngnick): Check if this is required.
+	// Collect all delegations for this secret.
 	for _, d := range kc.tlscertificatedelegations {
 		for _, cd := range d.Spec.Delegations {
-			for _, n := range cd.TargetNamespaces {
-				delegations[n+"/"+cd.SecretName] = true
+			if secretObj.Namespace == d.Namespace && secretObj.Name == cd.SecretName {
+				for _, namespace := range cd.TargetNamespaces {
+					if namespace == "*" {
+						// No need to check other delegations since any namespace is allowed.
+						delegatedForAnyNamespace = true
+						break
+					}
+					delegations = append(delegations, namespace)
+				}
 			}
 		}
 	}
 
-	for _, ingress := range kc.ingresses {
-		if ingress.Namespace == secret.Namespace {
-			for _, tls := range ingress.Spec.TLS {
-				if tls.SecretName == secret.Name {
-					return true
-				}
-			}
+	// Compare returns true if the changedSecret is the same as the referredSecret and delegation is permitted.
+	compare := func(changedSecret, referredSecret types.NamespacedName, namespaceOfReferrer string) bool {
+		// If the secret is in the same namespace as the referring object, then compare the names.
+		if changedSecret.Namespace == namespaceOfReferrer && changedSecret == referredSecret {
+			return true
 		}
-		if delegations[ingress.Namespace+"/"+secret.Name] {
-			for _, tls := range ingress.Spec.TLS {
-				if tls.SecretName == secret.Namespace+"/"+secret.Name {
-					return true
-				}
+
+		// If the secret is delegated to any namespace, then compare the names.
+		if delegatedForAnyNamespace && changedSecret == referredSecret {
+			return true
+		}
+
+		// If the secret is delegated to the namespace of the referring object, then compare the names.
+		for _, namespace := range delegations {
+			if namespace == namespaceOfReferrer && changedSecret == referredSecret {
+				return true
 			}
 		}
 
-		if delegations["*/"+secret.Name] {
-			for _, tls := range ingress.Spec.TLS {
-				if tls.SecretName == secret.Namespace+"/"+secret.Name {
-					return true
-				}
+		return false
+	}
+
+	for _, ingress := range kc.ingresses {
+		for _, tls := range ingress.Spec.TLS {
+			if compare(secret, k8s.NamespacedNameFrom(tls.SecretName, k8s.DefaultNamespace(ingress.Namespace)), ingress.Namespace) {
+				return true
```

Comment:
Two thoughts here:
- When we're checking each Ingress, HTTPProxy, etc. to see if it references `secretObj`, the _first_ thing we should do is see if the namespace/name of the reference match the namespace/name of `secretObj` (taking into account defaulting of the reference's namespace to the Ingress/HTTPProxy's namespace if it's not explicitly specified). If not, then the Ingress/HTTPProxy does not reference the Secret in question, so we don't have to do any further checks for delegation/etc. That'll be the most common case (most Ingresses/Proxies won't have a reference to the Secret in question), so we'll want to exit early if it's hit. 
- For checking if a reference is allowed, I think we should be able to use the existing `delegationPermitted` function here; it'd be nice to only have one (well-tested) version of that logic. Since we'll only hit this branch for Ingresses/HTTPProxies that _do_ reference the Secret, I'm not worried about small differences in performance.

So, for Ingress for example, I'm thinking something like:

```go
for _, ingress := range kc.ingresses {
	for _, tls := range ingress.Spec.TLS {
		// Get a NamespacedName for the Secret that the Ingress references, to make comparisons easier.
		ingressSecretRef := k8s.NamespacedNameFrom(tls.SecretName, k8s.DefaultNamespace(ingress.Namespace))

		// Secret is not referenced by this Ingress TLS: continue.
		if ingressSecretRef != secret {
			continue
		}

		// Check if Secret is allowed to be referenced by this Ingress
		// (this handles both same-namespace and cross-namespace cases).
		if kc.delegationPermitted(secret, ingress.Namespace) {
			return true
		}
	}
}```

Also note https://github.com/projectcontour/contour/issues/5538 which I noticed while looking at this; we can fix it separately.

Reply:
I agree it would be best to reuse `delegationPermitted()` but I was concerned it was very sub-optimal performance-wise for this particular use case: it would repeatedly try to find the same secret from all delegations, as many times as we have ingresses/HTTPProxies. Comparing the names first makes the call less frequent (assuming delegation is used rarely, which is likely the case) so maybe the performance does not matter?  

I used reverse order since as generally permission check would be nice to do first, but it does not matter in this case. However, preferably, I would like to combine the name comparison with permission check in a single function call, so that it will become less likely that any future edits will miss the permission check.

As last point, as I mentioned, I was also thinking of simplifying this all and completely removing the delegation check. Little harm would be done by extra DAG rebuild, even if it would happen for rare references that were invalid because of missing delegation?

Reply:
> As last point, as I mentioned, I was also thinking of simplifying this all and completely removing the delegation check. Little harm would be done by extra DAG rebuild, even if it would happen for rare references that were invalid because of missing delegation?

That's fair, I think that would be perfectly fine too and would avoid all of this!

Reply:
I went ahead changed the PR and removed the delegation checks. I changed the test cases a bit, removing some delegation checks from `TestSecretTriggersRebuild` which do not anymore have much meaning, but I kept the `TLSCertificateDelegation` objects in that test case for completeness, even though they are now ignored.